prov/util/src/util_mr_cache.c

```diff
@@ -478,7 +478,8 @@ void ofi_mr_cache_cleanup(struct ofi_mr_cache *cache)
 	ofi_rbmap_cleanup(&cache->tree);
 	if (cache->domain)
 		ofi_atomic_dec32(&cache->domain->ref);
-	ofi_bufpool_destroy(cache->entry_pool);
+	if (cache->entry_pool)
+		ofi_bufpool_destroy(cache->entry_pool);
```

Comment:
Is the code calling ofi_mr_cache_cleanup() after ofi_mr_cache_init() failed?  If so, that's not supported by the code and would need some other check to handle that case completely.

Reply:
There's a regression between v1.18.1 and main. Previously when `ofi_mr_cache_init()` failed, its `domain` field was set to NULL so that `ofi_mr_cache_cleanup()` could return-fast based on nullness of `cache->domain`:

5eaaf0677174788394ecbef22f84015eaebd9965 commit changed that and now `cache->domain` is not `NULL`ed in case of an error during `ofi_mr_cache_init()`.

The new code is wrong - it skips the `ofi_mr_cache_cleanup()` if the `cache->domain` is `NULL` which now doesn't mean that initialization failed, only that domain wasn't provided to `ofi_mr_cache_init()`. Because of that we lost the ability to distinguish between initialized MR cache with its domain set to NULL and MR cache that failed an initialization.